### PR TITLE
Do not force division to happen in `float`

### PR DESCRIPTION
**Pull request by eric-wieser**
This should make it possible to use `Clock()` on `Decimal` types.

The `frequency` attribute will still be computed as a `float`, but this isn't used for anything anyway.

These `.0` suffixes were needed in Python 2, but have little place in Python 3 code.

### CODE REVIEW

**Review by imphil — DISMISSED** [review]
LGTM once you also import Union :-)

**Review by eric-wieser — COMMENTED** [review]
Probably this should have a test of `Clock(Decimal("0.1"), "us")` or something before we actually merge. I don't know when I will get around to adding one.

**Review by ktbarrett — APPROVED** [review]

**Comment by ktbarrett** [comment]
Re-running.

File: documentation/source/newsfragments/3045.feature.rst
Comment (imphil) [inline documentation/source/newsfragments/3045.feature.rst:1]:
```suggestion
Add support for :class:`fraction.Fraction` and :class:`decimal.Decimal` to the `` period`` argument of :class:`cocotb.clock.Clock`.
```

**Review by imphil — APPROVED** [review]
LGTM with some small comments. Thanks @ktbarrett and @eric-wieser!

File: tests/test_cases/test_cocotb/test_clock.py
Comment (ktbarrett) [inline tests/test_cases/test_cocotb/test_clock.py:76]:
```suggestion
    with pytest.warns(FutureWarning, match='cause a CancelledError to be thrown'):
        clk1.cancel()
```

File: tests/test_cases/test_cocotb/test_clock.py
Comment (ktbarrett) [inline tests/test_cases/test_cocotb/test_clock.py:76]:
```suggestion
        clk2.cancel()
```

**Comment by imphil** [comment]
Squashed and updated commit message, ready for merge once CI passes.

**Comment by ktbarrett** [comment]
Trying from the start...

**Comment by imphil** [comment]
That was quick :-)

**Pull request merged**
